File: day_6/puzzle_1/solve.ts
import fs from "fs";
import { benchmark } from "../../benchmark/benchmark";

benchmark(() => {
  const file = fs.readFileSync("./day_6/input.txt", { encoding: "utf-8" });

  let startOfPacketMarker = -1;

  for (let i = 0; i < file.length; i++) {
    const packet = file.slice(i, i + 4);
    const checked: string[] = [];

    for (const value of packet) {
      if (checked.includes(value)) {
        break;
      } else {
        checked.push(value);
      }
    }

    if (checked.length === 4) {
      startOfPacketMarker = i + 4;
      break;
    }
  }

  return startOfPacketMarker;
}, "day_6_puzzle_1");

File: day_6/puzzle_1/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run('day_6_puzzle_1') do
  File.open('./day_6/input.txt', 'r') do |file|
    packet_stream = file.read

    start_of_packet_marker = -1

    packet_stream.length.times do |i|
      packet  = packet_stream[i..i + 3]
      checked = []

      packet.each_char do |value|
        break if checked.include?(value)

        checked.push(value)
      end

      if checked.length == 4
        start_of_packet_marker = i + 4
        break
      end
    end

    start_of_packet_marker
  end
end
